docs/contributing/engineering-values.md

```diff
@@ -0,0 +1,73 @@
+
+## What this is and what this isn't
+
+These values are not unbendable rules. They're an attempt to share knowledge about how the application came to be and how we currently think it should be developed. It's a non-exhaustive, living document of intentions, and beliefs.
+
+We've chosen to focus exclusively on the engineering (i.e. how we write code) part in this document. There are other values surrounding visual design, product and project management, community etc that we're not going to consider here.
+
+## Our values
+
+1. [Types are good](#types-are-good)
+2. [Immutability is good](#immutability-is-good)
+3. [Passing values to functions is good](#passing-values-to-functions-is-good)
+
+### Types are good
+
+We believe in strongly typed programming languages that help us avoid mistakes that are all too easy to make. When the compiler can't help us we believe in augmenting with other kinds of static analysis. Type systems and static analysis helps us not only in the moment we're writing code but also extends further out. By codifying our intentions we help prevent the next person from making a mistake.
+
+#### Examples
+
+Our [assertNever](https://github.com/desktop/desktop/blob/d26fd1ee670dfa7f16ded74b7a4108d2bfe68c79/app/src/lib/fatal-error.ts#L6-L21) helper lets us leverage the type system to verify exhaustiveness and get [compile-time errors](https://github.com/desktop/desktop/blob/8fc8e6f5d1a8153cc92bb0e324b9c26602211646/app/src/ui/branches/ci-status.tsx#L36-L47) when that assertion fails.
+
+Our [react-readonly-props-and-state](https://github.com/desktop/desktop/blob/d26fd1ee670dfa7f16ded74b7a4108d2bfe68c79/tslint-rules/reactReadonlyPropsAndStateRule.ts) static analysis ensures that we don't accidentally mutate state which React prohibits being mutated but isn't able to enforce due to the dynamic runtime.
+
+We [write our own type definitions](https://github.com/desktop/desktop/blob/eee92a96943afbc39057b1aae66c642e23dbf136/app/src/lib/globals.d.ts#L94-L112) when none exist.
+
+### Immutability is good
+
+By choosing to keep as much of our state as possible in immutable data structures we can be explicit about when and where we update it and have confidence that it's not going to change from underneath us. The last part is especially important in an application such as GitHub Desktop which has lots of state. The best way we've found so far to maintain our sanity is to be explicit about how that state flows through the app and when it's updated and immutability is one tool to help us stay on the right track.
+
+#### Examples
+
+We use [read-only versions of arrays](https://github.com/desktop/desktop/blob/a61a5bdc94ee8237dfff328957cdaee99a9b61e1/app/src/models/commit.ts#L21) in interfaces and object as well as [in function parameters](https://github.com/desktop/desktop/blob/355f9671860e4777827912ddc6aac44399f5732f/app/src/lib/email.ts#L17). (
```

Comment:
> We use read-only versions of arrays

How about an ESLint rule to enforce this, combined with a `MutableArray` type to make it easy to find mutable arrays?

<details><summary>What I’ve got so far</summary>

```js
context => {
  const sc = context.getSourceCode()
  function report(
    node,
    message = 'Don’t use `Array`. Use `ReadonlyArray` instead',
    fix = fixer => fixer.replaceText(node, 'ReadonlyArray')
  ) {
    context.report({
      node,
      message,
      fix
    })
  }
  function checkTypeAnnotation(ann) {
    if (ann.type === 'Identifier') {
      if (ann.name === 'Array') {
        report(ann)
      }
    } else if (ann.type === 'ArrayTypeAnnotation') {
      report(
        ann,
        `Don’t use \`${
        	sc.getText(ann)
        }\`. Use \`ReadonlyArray<${
        	sc.getText(ann.elementType)
        }>\` instead`,
        fixer => fixer.replaceText(
          ann,
          `ReadonlyArray<${sc.getText(ann.elementType)}>`
        )
      )
      checkTypeAnnotation(ann.elementType)
    } else if (
      ann.type === 'UnionTypeAnnotation' ||
      ann.type === 'IntersectionTypeAnnotation'
    ) {
      ann.types.forEach(checkTypeAnnotation)
    } else if (
      ann.type === 'GenericTypeAnnotation' ||
      ann.type === 'TSTypeAnnotation'
    ) {
      if (ann.typeParameters) {
        ann.typeParameters.params.forEach(checkTypeAnnotation)
      }
      checkTypeAnnotation(ann.id || ann.typeName)
    } else {
      context.report({
        node: ann,
        message: ann.type
      })
    }
  }
  return {
    Identifier(node) {
      if (node.typeAnnotation) {
          checkTypeAnnotation(node.typeAnnotation.typeAnnotation)
      }
    },
    TypeAnnotation(node) {
      if (node.typeAnnotation) {
        checkTypeAnnotation(node.typeAnnotation);
      }
    },
    'FunctionDeclaration, FunctionExpression, ArrowFunctionExpression'(node) {
      if (node.returnType) {
        checkTypeAnnotation(node.returnType.typeAnnotation)
      }
    }
  }
}
```

</details>